Miniscript support in Output Descriptors

darosior opened the pull request:
This adds Miniscript support for Output Descriptors without any signing logic (yet). See the OP of #24147 for a description of Miniscript and a rationale of having it in Bitcoin Core.
On its own, this PR adds "watchonly" support for Miniscript descriptors in the descriptor wallet. A follow-up adds signing support.

A minified corpus of Miniscript Descriptors for the `descriptor_parse` fuzz target is available at https://github.com/bitcoin-core/qa-assets/pull/92.
The Miniscript descriptors used in the unit tests here and in #24149 were cross-tested against the Rust implementation at https://github.com/rust-bitcoin/rust-miniscript.

This PR contains code and insights from Pieter Wuille.

darosior committed:
miniscript: add a helper to find the first insane sub with no child

This is helpful for finer grained descriptor parsing error: when there
are multiple errors to report in a Miniscript descriptor start with the
"smallest" fragments: the ones closer to be a leaf.

Co-Authored-By: Pieter Wuille <pieter@wuille.net>

darosior committed:
qa: better error reporting on descriptor parsing error

A nit, but was helpful when writing unit tests for Miniscript parsing

darosior committed:
miniscript: don't check for top level validity at parsing time

Letting the caller perform the checks allows for finer-grained error
reporting.

darosior committed:
qa: functional test Miniscript watchonly support

darosior commented:
Rebased.

darosior commented:
I (slightly) messed up the last rebase here, but there was some updates in the meantime so i'll push a fix along with those soon.

darosior committed:
Miniscript support in output descriptors

Miniscript descriptors are defined under P2WSH context (either `wsh()`
or `sh(wsh())`).
Only sane Miniscripts are accepted, as insane ones (although valid by
type) can have surprising behaviour with regard to malleability
guarantees and resources limitations.
As Miniscript descriptors are longer and more complex than "legacy"
descriptors, care was taken in error reporting to help a user determine
for what reason a provided Miniscript is insane.

Co-authored-by: Pieter Wuille <pieter.wuille@gmail.com>

darosior commented:
Rebased on top of #24860. I've made some cleanups and added more comments. I've also merged the unit tests of the Miniscript descriptors with the commit introducing the feature. I've changed the author of the commit to myself from Pieter as it was modified quite a lot from [his original one](https://github.com/bitcoin/bitcoin/pull/16800/commits/037e55adc005e9d8253d3d35e7a30b2e8521389c)  and i don't want to wrongly attribute him my own mistakes :p (of course, the co-author credits were switched accordingly).

darosior commented:
Hmm, so the use of `uint32_t` as key types makes the check for duplicate keys void. I need to look into it.

sipa commented:
@darosior Perhaps add an KeyCompare(Key, Key) function to contexts, and use that as set ordering in the duplicate check, rather than Key::operator< ?

OliverOffing commented on src/script/miniscript.h:
```diff
@@ -298,6 +298,8 @@ struct Node {
     //! Cached script length (computed by CalcScriptLen).
     const size_t scriptlen;
+    //! Whether a public key appears more than once in this node.
+    const bool duplicate_key;
```
Naming could be improved. The name `duplicate_key` might suggest that the variable holds the value or reference of the key that is duplicated, when in fact it can only hold either `true` or `false`—and there could be multiple duplicated keys. It'd be best to name this variable `has_duplicate_keys` instead, which would go nicely along the naming scheme of `IsNonMalleable` and `IsSane`.

darosior replied:
> The name duplicate_key might suggest that the variable holds the value or reference of the key that is duplicated, when in fact it can only hold either true or false

I don't think it's worth making the variable name even longer. The `bool` type should be enough for one to get that this value holds, well, a boolean and not a reference to the key.
Also note this commit is part of https://github.com/bitcoin/bitcoin/pull/24860 which features a `ContainsDuplicateKey()` public API in the same spirit as `IsNonMalleable` and `IsSane`.

> and there could be multiple duplicated keys

The detection algorithm cuts through and wouldn't know that, it just knows that "there exists one duplicate key".

darosior commented:
Rebased on top of #24860 to integrate the `IsSane` renaming and key duplicate check using a `KeyCompare` method. I added a unit test for duplicate key check when parsing a Miniscript descriptor and took the opportunity to rework a bit the error messages to make them consistent (Now always "is not sane: [reason]").

sipa commented on src/script/descriptor.cpp:
```diff
@@ -917,6 +918,89 @@ class TRDescriptor final : public DescriptorImpl
+    const std::vector<CPubKey>& m_keys;
+
+public:
+    ScriptMaker(const std::vector<CPubKey>& keys) : m_keys(keys) {}
```
Add LIFETIMEBOUND attribute to `keys`?

sipa commented on src/script/descriptor.cpp:
```diff
@@ -917,6 +918,89 @@ class TRDescriptor final : public DescriptorImpl
+    bool m_private;
+
+public:
+    StringMaker(const SigningProvider* arg, const std::vector<std::unique_ptr<PubkeyProvider>>& pubkeys, bool priv)
```
Add LIFETIMEBOUND attribute to `arg` and to `pubkeys`?

sipa commented on src/script/descriptor.cpp:
```diff
@@ -1058,6 +1142,93 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(uint32_t key_exp_index, const Span<c
+    //! Used to detect key parsing errors within a Miniscript.
+    mutable std::string m_key_parsing_error;
+
+    KeyParser(FlatSigningProvider* out, const SigningProvider* in) : m_out(out), m_in(in) {}
```
Add LIFETIMEBOUND here to `out` and `in` as well?

sipa commented on src/script/descriptor.cpp:
```diff
@@ -1058,6 +1142,93 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(uint32_t key_exp_index, const Span<c
+    KeyParser(FlatSigningProvider* out, const SigningProvider* in) : m_out(out), m_in(in) {}
+
+    bool KeyCompare(const Key& a, const Key& b) const {
+        return m_keys[a]->ToString() < m_keys[b]->ToString();
```
That sounds pretty slow. Perhaps it's worth adding generic comparison operators to the PubkeyProvider classes? Could be left as a TODO.

darosior replied:
Yeah, it's also slightly incorrect. For instance if a user -for whatever reason- has a descriptor with 2 xpubs with the same key but different chaincodes, this would not detect them as duplicates (same with same key but diff origins). I wanted to revisit that but it slipped my mind.

I added an `operator<` to `PubkeyProvider`:
```patch
diff --git a/src/script/descriptor.cpp b/src/script/descriptor.cpp
index 94c9071999..66c4e1cba9 100644
--- a/src/script/descriptor.cpp
+++ b/src/script/descriptor.cpp
@@ -162,6 +162,20 @@ public:
 
     virtual ~PubkeyProvider() = default;
 
+    /** Compare two public keys represented by this provider.
+     * Used by the Miniscript descriptors to check for duplicate keys in the script.
+     */
+    bool operator<(PubkeyProvider& other) const {
+        CPubKey a, b;
+        SigningProvider dummy;
+        KeyOriginInfo dummy_info;
+
+        GetPubKey(0, dummy, a, dummy_info);
+        other.GetPubKey(0, dummy, b, dummy_info);
+
+        return a < b;
+    }
+
     /** Derive a public key.
      *  read_cache is the cache to read keys from (if not nullptr)
      *  write_cache is the cache to write keys to (if not nullptr)
@@ -1183,7 +1197,7 @@ struct KeyParser {
     KeyParser(FlatSigningProvider* out LIFETIMEBOUND, const SigningProvider* in LIFETIMEBOUND) : m_out(out), m_in(in) {}
 
     bool KeyCompare(const Key& a, const Key& b) const {
-        return m_keys[a]->ToString() < m_keys[b]->ToString();
+        return *m_keys[a] < *m_keys[b];
     }
 
     template<typename I> std::optional<Key> FromString(I begin, I end) const
```

sipa commented on src/script/descriptor.cpp:
```diff
@@ -1058,6 +1142,93 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(uint32_t key_exp_index, const Span<c
+    {
+        assert(m_out);
+        Key key = m_keys.size();
+        auto pk = ParsePubkey(key, Span<const char>(&*begin, &*end), ParseScriptContext::P2WSH, *m_out, m_key_parsing_error);
```
I think `{&*begin, &*end}` can be used instead of `Span<const char>(&*begin, &*end)`.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -1058,6 +1142,93 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(uint32_t key_exp_index, const Span<c
+    {
+        assert(end - begin == 20);
+        assert(m_in);
+        uint160 hash(std::vector<unsigned char>(begin, end));
```
Constructing a vector just to turn it into a `uint160` sounds a bit inefficient. I think this works too:

```c++
uint160 hash;
std::copy(begin, end, hash.begin());`
```

sipa commented on src/script/descriptor.cpp:
```diff
@@ -1279,6 +1450,38 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t& key_exp_index, Span<const
+    // Process miniscript expressions.
+    {
+        KeyParser parser(&out, nullptr);
+        auto node = miniscript::FromString(std::string(expr.begin(), expr.end()), parser);
```
Perhaps for a follow-up PR, but maybe we want to change the `FromString` methods (in miniscript code, but also descriptors in general)` to take `std::string_view`s? That would avoid intermediary string objects like here.

darosior replied:
Since it's going to touch the Miniscript code again, i'd prefer to keep this as a follow-up if you don't mind.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -1279,6 +1450,38 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t& key_exp_index, Span<const
+                else {
+                    error += " is not sane";
+                    if (!insane_node->IsNonMalleable()) error += ": malleable witnesses exist";
+                    else if (insane_node == node.get() && !insane_node->NeedsSignature()) error += ": witnesses without signature exist";
```
Use braces and separate lines, unless for single-line "then" statements without "else"s.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -493,12 +508,12 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+    //! Public key arguments for this descriptor (size 1 for PK, PKH, WPKH; any size for WSH and Multisig).
     const std::vector<std::unique_ptr<PubkeyProvider>> m_pubkey_args;
     //! The string name of the descriptor function.
     const std::string m_name;
```
In e9cac52614d11e30d1f6fec33efce3d00953bd70 "Miniscript support in output descriptors"

nit: `m_name` does not need to be `protected`.

darosior replied:
Yeah, but 
```patch
diff --git a/src/script/descriptor.cpp b/src/script/descriptor.cpp
index 66c4e1cba9..b40e330f67 100644
--- a/src/script/descriptor.cpp
+++ b/src/script/descriptor.cpp
@@ -508,11 +508,12 @@ public:
 /** Base class for all Descriptor implementations. */
 class DescriptorImpl : public Descriptor
 {
+    //! The string name of the descriptor function.
+    const std::string m_name;
+
 protected:
     //! Public key arguments for this descriptor (size 1 for PK, PKH, WPKH; any size for WSH and Multisig).
     const std::vector<std::unique_ptr<PubkeyProvider>> m_pubkey_args;
-    //! The string name of the descriptor function.
-    const std::string m_name;
 
     //! The sub-descriptor arguments (empty for everything but SH and WSH).
     //! In doc/descriptors.m this is referred to as SCRIPT expressions sh(SCRIPT)
```

Would make me reorder the initializer list of all classes heriting from `DescriptorImpl`, so i guess it's not so bad if it's just `protected` instead of `private`? Happy to change if you prefer, still. (This or doing `protectred:` then `private:` then `protected:` again.)

achow101 reviewed (commented):
4665124eca32c654df55fccf314f74d8fb9731be "miniscript: add a helper to find the deepest insane sub" does not seem to find the deepest, just an insane sub that has no children?

Why not have `Key` be `PubkeyProvider` so that each miniscript node can handle its own keys?

darosior commented:
@sipa: could you give this another look when you have a minute?

darosior commented:
I've got a decent fuzzing coverage of this PR with the corpus at https://github.com/bitcoin-core/qa-assets/pull/92 (using this branch along with #25540 to avoid memory blowups). A coverage report is available [here](http://download.darosior.ninja/dump/coverage_descriptor_parse_miniscript_2205/src/script/index.html).

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -917,6 +932,89 @@ class TRDescriptor final : public DescriptorImpl
+    {
+        std::string ret;
+        if (m_private) {
+            if (!m_pubkeys[key]->ToPrivateString(*m_arg, ret)) return {};
```
What is supposed to happen here if we have a descriptor with 1 private key and otherwise only public keys?

darosior replied:
The same as for existing descriptors, a failure:
```diff
diff --git a/src/test/descriptor_tests.cpp b/src/test/descriptor_tests.cpp
index a8c666079d..668cddfbba 100644
--- a/src/test/descriptor_tests.cpp
+++ b/src/test/descriptor_tests.cpp
@@ -488,6 +488,13 @@ Check("sh(wsh(multi(20,KzoAz5CanayRKex3fSLQ2BwJpN7U52gZvxMyk78nDMHuqrUxuSJy,KwGN
 
     // Miniscript tests
 
+    FlatSigningProvider prov;
+    std::string err, out;
+    const auto multi_desc = Parse("wsh(multi(1,L4gM1FBdyHNpkzsFh9ipnofLhpZRp2mwobpeULy1a6dBTvw8Ywtd,032707170c71d8f75e4ca4e3fce870b9409dcaf12b051d3bcadff74747fa7619c0))", prov, err, false);
+    assert(multi_desc && !multi_desc->ToPrivateString(prov, out));
+    const auto ms_desc = Parse("wsh(or_i(pk(L4gM1FBdyHNpkzsFh9ipnofLhpZRp2mwobpeULy1a6dBTvw8Ywtd),pk(032707170c71d8f75e4ca4e3fce870b9409dcaf12b051d3bcadff74747fa7619c0)))", prov, err, false);
+    assert(ms_desc && !ms_desc->ToPrivateString(prov, out));
+
     // Invalid checksum
     CheckUnparsable("wsh(and_v(vc:andor(pk(L4gM1FBdyHNpkzsFh9ipnofLhpZRp2mwobpeULy1a6dBTvw8Ywtd),pk_k(Kx9HCDjGiwFcgVNhTrS5z5NeZdD6veeam61eDxLDCkGWujvL4Gnn),and_v(v:older(1),pk_k(L4o2kDvXXDRH2VS9uBnouScLduWt4dZnM25se7kvEjJeQ285en2A))),after(10)))#abcdef12", "wsh(and_v(vc:andor(pk(03cdabb7f2dce7bfbd8a0b9570c6fd1e712e5d64045e9d6b517b3d5072251dc204),pk_k(032707170c71d8f75e4ca4e3fce870b9409dcaf12b051d3bcadff74747fa7619c0),and_v(v:older(1),pk_k(02aa27e5eb2c185e87cd1dbc3e0efc9cb1175235e0259df1713424941c3cb40402))),after(10)))#abcdef12", "Provided checksum 'abcdef12' does not match computed checksum 'tyzp6a7p'");
     // Only p2wsh context is valid
```

It is certainly suboptimal for certain usecases (like importing a partially private desc and calling `listdescriptors true`), but it's the existing behaviour and i don't think it should be fixed in this PR.

sipa replied:
See #24361

darosior commented:
Indeed a test for `FindInsaneSub` is redundant with the descriptor parsing failure test vectors. But it's trivial so i added one. I also took this opportunity to rename the commit to not mention it's looking for the "deepest" insane sub, as noted by @achow101.

`InferPubkey` is used in the `KeyParser` so it needs to be declared. I made `InferXOnlyPubkey` stick with the former for pure style/organisation reasons: it appeared to be "righter" and not harder to review (this just extends the moved patch from ~8 to ~16 lines which is trivially reviewable with `--color-moved` as you mentioned).

As for splitting the commit, i'd prefer to leave it like that if you don't feel too strongly about it (and since you marked it as nit). I think it would just make the commits less atomic, and the `StringMaker`/`ScriptMaker` are tested through the `MiniscriptDescriptor`.

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -1058,6 +1156,94 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(uint32_t key_exp_index, const Span<c
+    KeyParser(FlatSigningProvider* out LIFETIMEBOUND, const SigningProvider* in LIFETIMEBOUND) : m_out(out), m_in(in) {}
+
+    bool KeyCompare(const Key& a, const Key& b) const {
+        return *m_keys[a] < *m_keys[b];
```
Maybe use `.at(a)` here, as well as in `ToString()`

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -1279,6 +1465,44 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t& key_exp_index, Span<const
+                auto insane_node = node.get();
+                if (const auto sub = node->FindInsaneSub()) insane_node = sub;
+                if (const auto str = insane_node->ToString(parser)) error = *str;
+                if (!insane_node->IsValid()) { error += " is invalid";}
```
```cpp
if (!insane_node->IsValid()) {
  error += " is invalid";
} else {
```

Sjors reviewed (approved):
tACK 49ac002733157f5629303d1cf59fa3151fdf1e73

I was able to import a miniscript descriptor into a watch-only wallet and send some signet coins to it. I'll continue testing after #24148 is rebased. 

Nit: when you use an invalid miniscript fragment, like `or_v(…)` it says "A function is needed within P2WSH". It might be more clear to use `or_v(…) is invalid` here.

achow101 commented:
re-ACK 49ac002733157f5629303d1cf59fa3151fdf1e73

instagibbs commented on test/functional/wallet_miniscript.py:
```diff
@@ -0,0 +1,93 @@
+        )
+        self.ms_wo_wallet = self.nodes[0].get_wallet_rpc("ms_wo")
+
+        # Sanity check we wouldn't let an insane Miniscript descriptor in
```
need to define insane(I don't know what that means)

darosior replied:
It was discussed in the OP of the Miniscript PR, #24147, and is defined in `src/script/miniscript.h`:
https://github.com/bitcoin/bitcoin/blob/062b9db0ccb6af8bfbaa2b29132408cda9991b40/src/script/miniscript.h#L842-L846

w0xlt commented on src/script/descriptor.cpp:
```diff
@@ -1058,6 +1156,94 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(uint32_t key_exp_index, const Span<c
+        return m_keys[key]->ToString();
+    }
+
+    template<typename I> std::optional<Key> FromPKBytes(I begin, I end) const
```
nit: Maybe is `FromPublicKeyBytes()` clearer ?

darosior replied:
This interface was introduced in #24147, i'm not going to change it and all the callsites now. :)

w0xlt commented on src/script/descriptor.cpp:
```diff
@@ -1058,6 +1156,94 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(uint32_t key_exp_index, const Span<c
+        return {};
+    }
+
+    template<typename I> std::optional<Key> FromPKHBytes(I begin, I end) const
```
nit: Maybe is `FromPublicKeyHashBytes()` clearer ?

darosior replied:
Same here

w0xlt reviewed (approved):
Tested ACK https://github.com/bitcoin/bitcoin/pull/24148/commits/49ac002733157f5629303d1cf59fa3151fdf1e73

The first commit (https://github.com/bitcoin/bitcoin/pull/24148/commits/d4b6f6842a5273afd1e7fe68d18d272566eb90c8) fails to run the miniscript unit test.

```
$ git reset --hard d4b6f6842a5273afd1e7fe68d18d272566eb90c8
$ make
$ src/test/test_bitcoin --run_test=miniscript_tests 
Running 1 test case...
test/miniscript_tests.cpp(303): error: in "miniscript_tests/fixed_tests": check ms_ins && !ms_ins->IsSane() has failed
unknown location(0): fatal error: in "miniscript_tests/fixed_tests": memory access violation at address: 0x00000038: no mapping at fault address
test/miniscript_tests.cpp(303): last checkpoint
```

The second commit (https://github.com/bitcoin/bitcoin/pull/24148/commits/10b905fe98ccef3840e81664ea36fe11860f1d6d) fixes this error.

Maybe they might be squashed or the test file changed in the second.

darosior commented:
Thanks @w0xlt, good catch: it's the unit test for `FindInsaneSub` i just added that i had only tested on the tip of the branch. I've swapped the two first commits now and they all individually compile and pass the tests.

@Sjors: i was missing an `a:` wrapper for the second `after`, i've added a check that the Miniscript is valid (which does not change the test since an invalid sub is necessarily insane):
```diff
diff --git a/src/test/miniscript_tests.cpp b/src/test/miniscript_tests.cpp
index f5bea00e1e..bc5c49ef63 100644
--- a/src/test/miniscript_tests.cpp
+++ b/src/test/miniscript_tests.cpp
@@ -299,10 +299,10 @@ BOOST_AUTO_TEST_CASE(fixed_tests)
     // 2. It contains timelock mixes
     // We'll report the timelock mix error, as it's "deeper" (closer to be a leaf node) than the "no 's' property"
     // error is.
-    const auto ms_ins = miniscript::FromString("or_i(and_b(after(1),after(10000000)),pk(03cdabb7f2dce7bfbd8a0b9570c6fd1e712e5d64045e9d6b517b3d5072251dc204))", CONVERTER);
-    BOOST_CHECK(ms_ins && !ms_ins->IsSane());
+    const auto ms_ins = miniscript::FromString("or_i(and_b(after(1),a:after(1000000000)),pk(03cdabb7f2dce7bfbd8a0b9570c6fd1e712e5d64045e9d6b517b3d5072251dc204))", CONVERTER);
+    BOOST_CHECK(ms_ins && ms_ins->IsValid() && !ms_ins->IsSane());
     const auto insane_sub = ms_ins->FindInsaneSub();
-    BOOST_CHECK(insane_sub && *insane_sub->ToString(CONVERTER) == "and_b(after(1),after(10000000))");
+    BOOST_CHECK(insane_sub && *insane_sub->ToString(CONVERTER) == "and_b(after(1),a:after(1000000000))");
 
     // Timelock tests
     Test("after(100)", "?", TESTMODE_VALID | TESTMODE_NONMAL); // only heightlock
```

I also took the opportunity to fix the two nits @Sjors had previously.

w0xlt reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/24148/commits/ffc79b8e492c6dd1352e528fd82e45d8d25eaa04

instagibbs reviewed (approved):
only reviewed the tests and interface, non blocking nits only

achow101 commented:
ACK ffc79b8e492c6dd1352e528fd82e45d8d25eaa04

sanket1729 reviewed (commented):
post-merge ACK. Did not review the tests. I like the error reporting by looking for inner insane sub.

Sjors commented:
Opening a miniscript-descriptor wallet in an older version of Bitcoin Core causes a crash. (How) did we handle backward compatibility for `tr()` descriptors? Do we want to set a flag each time? Of have some generic mechanism to abort loading a wallet that has unrecognised descriptor / miniscript fragment?

mzumsande commented on src/script/descriptor.cpp:
```diff
@@ -1058,6 +1156,94 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(uint32_t key_exp_index, const Span<c
+        assert(m_out);
+        Key key = m_keys.size();
+        auto pk = ParsePubkey(key, {&*begin, &*end}, ParseScriptContext::P2WSH, *m_out, m_key_parsing_error);
+        if (!pk) return {};
```
I get two compiler warnings on master from this, `gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0`:
```
script/descriptor.cpp: In function ‘miniscript::NodeRef<Key> miniscript::internal::DecodeScript(I&, I, const Ctx&) [with Key = unsigned int; Ctx = {anonymous}::KeyParser; I = __gnu_cxx::__normal_iterator<std::pair<opcodetype, std::vector<unsigned char> >*, std::vector<std::pair<opcodetype, std::vector<unsigned char> > > >]’:
script/descriptor.cpp:1243:17: warning: ‘<anonymous>’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 1243 |         return {};
      |                 ^
script/descriptor.cpp: In function ‘miniscript::NodeRef<Key> miniscript::internal::Parse(Span<const char>, const Ctx&) [with Key = unsigned int; Ctx = {anonymous}::KeyParser]’:
script/descriptor.cpp:1208:26: warning: ‘<anonymous>’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 1208 |         if (!pk) return {};
```

sipa replied:
I believe that's a known false positive in certain GCC versions.

raywhite714 replied:
I don't know what any of this means people I don't even know.Have access to my keys and addresses

raywhite714 commented:
I don't know what none of this means

On Tue, Jul 19, 2022, 4:11 PM Pieter Wuille ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/script/descriptor.cpp
> <https://github.com/bitcoin/bitcoin/pull/24148#discussion_r924960870>:
>
> > +    mutable std::vector<std::unique_ptr<PubkeyProvider>> m_keys;
> +    //! Used to detect key parsing errors within a Miniscript.
> +    mutable std::string m_key_parsing_error;
> +
> +    KeyParser(FlatSigningProvider* out LIFETIMEBOUND, const SigningProvider* in LIFETIMEBOUND) : m_out(out), m_in(in) {}
> +
> +    bool KeyCompare(const Key& a, const Key& b) const {
> +        return *m_keys.at(a) < *m_keys.at(b);
> +    }
> +
> +    template<typename I> std::optional<Key> FromString(I begin, I end) const
> +    {
> +        assert(m_out);
> +        Key key = m_keys.size();
> +        auto pk = ParsePubkey(key, {&*begin, &*end}, ParseScriptContext::P2WSH, *m_out, m_key_parsing_error);
> +        if (!pk) return {};
>
> I believe that's a known false positive in certain GCC versions.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/24148#discussion_r924960870>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ATA6EMVYKGXUOPGDAVBHS2TVU4KZRANCNFSM5MX3GFXA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

S3RK commented:
> Opening a miniscript-descriptor wallet in an older version of Bitcoin Core causes a crash. (How) did we handle backward compatibility for `tr()` descriptors? Do we want to set a flag each time? Of have some generic mechanism to abort loading a wallet that has unrecognised descriptor / miniscript fragment?

Not sure how it was done with `tr()` descriptors. But before `CWallet::SetMinVersion` was used to handle wallet compatibility with various software versions.
